### DISCUSSION TITLE
Save UI separate from logic

### DISCUSSION THREAD

**Discussion by LalFX** [TomSchimansky/CustomTkinter#1097] [General]
First, thank you for the implementation. Its one of the best looking GUI solutions I have found after looking really hard for something that looks modern.

Would it be possible to separate the UI code from the logic code, perhaps saving the UI in another py file and loading this into the main app? This would clean up the code a lot  as right now there is so much of UI code that lives in the main code loop?

I am relatively new to python - so perhaps there is an easy way to do this that I have missed. If so, I apologize and would appreciate any feedback you can provide.

**Comment by Kkleytt** [comment]
Конечно можно, просто раздели свой backend проект на логические команды, которые можно отделаить друг от друга(к примеру, ввод от пользователя 3 данных, месяц, день и год, чтобы высчитать сколько дней он прожил, и вызывай функцию пересчета только после того как все эти данные будут у программы(к примеру в блоке ввода, чтобы он был заполнен) и вызывай свой backend проект, для этого просто создай файл .py с функциями и вызови его как локальную библиотеку в файле с GUI, тогда вся логика будет выполнятся незаметно для пользователя через нужные для тебя функции. Надеюсь я дал подробный ответ на твой вопрос, если что пиши еще!

**Reply by LalFX** [reply]
Thanks - that was what i had mind. However, what i was hoping is to have the logic in the main py file and load the gui from an external py file. That way the logic file - which is the main file anyway does not get cluttered up with the GUI code.

**Comment by LalFX** [comment]
I sorted out a solution for anyone who was a newbie and is looking to solve this issue.

The reason I wanted to move the GUI code from the main code was mainly organizational as the GUI code can get very long and clutters up the logic and process code. I managed to solve this issue by creating a separate widgets.py file with a Class WidgetCreator. 
Something like this:

widget.py:

```
class WidgetCreator:
    def __init__(self, app, ctk, filedialog, PIL):
        self.app = app
        self.ctk = ctk
        self.filedialog = filedialog
        self.PIL = PIL

    def create_widgets(self):
        self.app.input_folder_path = self.ctk.StringVar()
        self.app.output_folder_path = self.ctk.StringVar()
        self.app.quality = self.ctk.IntVar(value=60)
```
Then I would call this class in the main_gui.py like below:
main_gui.py:

```
class ImageCompressorApp(ctk.CTk):
    def __init__(self):
        super().__init__()
        
        self.setup_window() # sets up the geometry of the app
        self.set_app_icon() # sets up the icons for the app

        # Create an instance of WidgetCreator and use it to create widgets
        widget_creator = WidgetCreator(self, ctk, filedialog, PIL)
        self.main_frame = widget_creator.create_widgets()
        
        # Bind events after widget creation
        self.bind_events()
```
And then load everything like normal:

```
if __name__ == "__main__":
    app = ImageCompressorApp()
    app.mainloop()
```

Hope this helps.
